src/reducers/index.js

```diff
@@ -7,6 +7,30 @@ const dbReducer = (state = {}, action) => {
         return _.extend({ [action.payload.id] : action.payload}, state)
       }
       break;
+    case 'OPTIMISTIC_UPDATE_LOCATION':
+      if(action.payload){
+        return _.extend(
+          {},
+          state,
+          {
+            [`last/${action.payload.id}`] : state[action.payload.id],
+            [action.payload.id] : _.extend({}, state[action.payload.id], action.payload.params)
+          }
+        )
+      }
+      break;
+    case 'ROLLBACK_UPDATE_LOCATION':
+      if(action.payload){
+        return _.extend(
+          {},
+          state,
+          {
+            [`last/${action.payload.id}`] : null,
+            [action.payload.id] : state[`last/${action.payload.id}`]
```

Comment:
Hmmm, implementing rollback in the reducer like this is interesting...
It certainly feels more Redux-y, though wouldn't this potentially lead to (probably rare) bugs when responses are slow, so the user can make more than 1 request before getting a response? (I mean, if e.g. 2 requests are made and then the first returns an error, it's actually the 2nd update that would get rolled back.)

Maybe it would be safer to leave rollback up to the thunk? The thunk could remember the exact fields that it changed and their old values, and in case of an error, dispatch an update with the old values for only those specific props.
Theoretically it could even remain the reducer's responsibility, but instead of keeping just 1 "last" value, it would store a mapping of request IDs to the changes they made.

Of course, even these approaches wouldn't help in the case both requests made changes to the same field.
Seems like handling that case would require maintaining a more complex history...

I would assume this is probably a common use case, so actually some source indicating what the best practice is would be great...
Googling briefly, it looks like https://github.com/ForbesLindesay/redux-optimist may have a request ID-based implementation ("transaction ID" technicallly) with a more complex history, so maybe that could work.

Reply:
Yeah, you're right. We probably shouldn't do the thunk option.
I think redux-optimistic is the way to go.

Reply:
I mean, thunks do have access to the state (they get passed both dispatch and getState), but the thunk approach still wouldn't solve the case where it's the same field being updated multiple times, and regardless I think it's just better to go with an established solution rather than build our own.